src/gadgets/k_mix.rs

```diff
@@ -1,61 +1,213 @@
 #![allow(non_snake_case)]
 
 use super::mix;
-use bulletproofs::r1cs::ConstraintSystem;
-use error::SpacesuitError;
+use bulletproofs::r1cs::{ConstraintSystem, R1CSError};
+use curve25519_dalek::scalar::Scalar;
 use std::iter::once;
-use value::AllocatedValue;
+use subtle::{ConditionallySelectable, ConstantTimeEq};
+use value::{AllocatedValue, Value};
 
 /// Enforces that the outputs are either a merge of the inputs: `D = A + B && C = 0`,
 /// or the outputs are equal to the inputs `C = A && D = B`. See spec for more details.
 /// Works for `k` inputs and `k` outputs.
 pub fn fill_cs<CS: ConstraintSystem>(
```

Comment:
I guess it's worth documenting that the arguments are provided in arbitrary order, and Result is the arrays of sorted gadget-inputs and gadget-outputs, such that the gadget creates constraints between them, but the user is also supposed to link gadget-inputs with their arguments through a shuffle.

Reply:
yes that was a stale comment, updated!